Fix Windows bootstrap panic on invalid symlink removal (issue #143045)

hasip-timurtas opened the pull request:
Fix Windows bootstrap panic on invalid symlink removal (issue rust-lang/rust#143045)

On Windows, use `fs::remove_file` for symlinks instead of `fs::remove_dir` to avoid error 267 when the symlink target is invalid. `fs::remove_file` can handle both file and directory symlinks/junctions, even when their targets don't exist.

r? @ghost

hasip-timurtas committed:
Fix typos and improve Windows symlink handling in `lib.rs`

- Corrected a typo in the error message from "Couldn’t" to "Couldn't".
- Enhanced the logic for removing symlinks on Windows to handle invalid targets more gracefully by attempting to remove them as files before falling back to directory removal.

workingjubilee commented:
@hasip-timurtas Please make your PR description concise so your reviewer does not waste time reading it. It should only need to be about two sentences for a change of this size.

workingjubilee commented:
Oh, your original commit message contains `r? @ghost`

That's something for guiding rustbot. You need to put it in your PR description for it to work. I can help, though, by unassigning your reviewer, since r? @ghost is for a PR that you do not intend to have reviewed.

hasip-timurtas commented:
@workingjubilee please check if it's ok now 🙂

ChrisDenton reviewed (commented):
The comment is incorrect and I also don't think the code is the correct fix. Both `remove_file` and `remove_dir` can remove symlinks. However, `remove_dir` can only remove directory symlinks.

hasip-timurtas committed:
Refactor Windows symlink removal logic in `lib.rs`

- Simplified the handling of symlinks on Windows by removing the fallback to directory removal. The `fs::remove_file` function now directly handles both file and directory symlinks, even when the target is invalid, improving code clarity and reliability.

hasip-timurtas commented:
@ChrisDenton  Thank you for the feedback

After digging into the Windows implementation, I found that `fs::remove_file` actually has special handling for this case. When it fails with `ERROR_ACCESS_DENIED`, it retries with `FILE_FLAG_OPEN_REPARSE_POINT` to handle the symlink directly (see `library/std/src/sys/fs/windows.rs:1228-1235`).

This is why it works on invalid symlinks - it doesn't try to follow them. `fs::remove_dir` doesn't have this fallback, which causes error 267.

Since `fs::remove_file` handles both file and directory symlinks on Windows, it's the right choice here. Happy to add a comment explaining this if you think it'd be helpful.

ChrisDenton commented:
> Since `fs::remove_file` handles both file and directory symlinks on Windows, it's the right choice here.

Sorry, I think you've misunderstood what that code is doing. Both `DeleteFile` and the fallback will open the symlink itself not the target. There is no issue if the target does not exist. The fallback is only there for readonly files that `DeleteFile` can't handle. The problem here is different.

On Windows there are two types of symlink: directory and non-directory. `remove_dir` can delete directory symlinks and `remove_file` can delete non-directory symlinks. Mixing them up will not work. Whether or not a symlink is a directory or non-directory symlink is a property of the symlink itself and nothing to do with the target.

So I would prefer this be implemented something like: if `remove_dir` fails with `ErrorKind::NotADirectory` then retry with `remove_file`.

hasip-timurtas committed:
Refactor Windows symlink removal logic in `lib.rs`

- Updated the symlink removal process on Windows to first attempt removing directory symlinks with `fs::remove_dir`, and fallback to `fs::remove_file` for file symlinks. This change improves error handling and clarity in the code.

hasip-timurtas commented:
@ChrisDenton  Thank you for the clarification. I've implemented your suggested approach, but I'd like to confirm a few details to ensure we're handling all scenarios correctly:

1. Will `fs::remove_dir` on a directory symlink with an invalid target still succeed, or could it also fail with error 267?

2. Are there any other error codes besides `NotADirectory` that should trigger the fallback to `remove_file`?

3. Should we handle the case where both `remove_dir` and `remove_file` fail differently, or is panicking the appropriate behavior?

The current implementation:
```rust
if let Err(e) = fs::remove_dir(&host) {
    if e.kind() == io::ErrorKind::NotADirectory {
        t!(fs::remove_file(&host));
    } else {
        panic!("failed to remove symlink: {}", e);
    }
}
```

I want to make sure we're covering all the edge cases for this Windows-specific issue.

rust-log-analyzer commented:
The job **`mingw-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15944095674/job/44975972654) [(plain)](https://github.com/rust-lang/rust/commit/14bfecb7b9e81ec65bb3b66fedf13a581e717973/checks/44975972654/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] build_helper test:false 0.671
error: variables can be used directly in the `format!` string
   --> src/bootstrap/src/lib.rs:552:21
    |
552 |                     panic!("failed to remove symlink: {}", e);
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args
    = note: `-D clippy::uninlined-format-args` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::uninlined_format_args)]`
help: change this to
    |
552 -                     panic!("failed to remove symlink: {}", e);
552 +                     panic!("failed to remove symlink: {e}");
    |

[RUSTC-TIMING] bootstrap test:false 9.418
error: could not compile `bootstrap` (lib) due to 1 previous error
Build completed unsuccessfully in 0:16:41
```

</details>

hasip-timurtas committed:
Update error message formatting in `lib.rs`

- Changed the error message in the symlink removal logic to use the more concise formatting syntax `{e}` instead of the longer `format!` style. This improves readability and maintains consistency in error handling.

jieyouxu commented:
Can this just use `build_helper::fs::recursive_remove`? I want to say that doesn't traverse symlinks.

Dylan-DPC commented:
@hasip-timurtas any updates on this? thanks
